src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/PreferAsSpanOverSubstringTests.cs

```diff
@@ -0,0 +1,1854 @@
+            {
+                TestState =
+                {
+                    Sources = { testCode },
+                    AdditionalProjects = { { receiverProject.Name, receiverProject } },
+                    AdditionalProjectReferences = { receiverProject.Name },
+                    ExpectedDiagnostics = { VB.DiagnosticAt(0) }
+                },
+                FixedState =
+                {
+                    Sources = { fixedCode },
+                    AdditionalProjects = { { receiverProject.Name, receiverProject } },
+                    AdditionalProjectReferences = { receiverProject.Name }
+                },
+                ReferenceAssemblies = ReferenceAssemblies.Net.Net50
+            };
+            return test.RunAsync();
+        }
+
+        public static IEnumerable<object[]> Data_NestedViolations
+        {
+            get
+            {
+                string receiver = CS.Usings + @"
+public class C
+{
+    public static string Fwd(string text) => throw null;
+    public static string Fwd(Roschar span) => throw null;
+    public static string Fwd(string text1, string text2) => throw null;
+    public static string Fwd(Roschar span1, Roschar span2) => throw null;
+
+    public static void Consume(string text) { }
+    public static void Consume(Roschar span) { }
+    public static void Consume(string text1, string text2) { }
+    public static void Consume(Roschar span1, Roschar span2) { }
+}";
+                yield return new object[]
+                {
+                    receiver,
+                    @"{|#0:C.Consume({|#1:C.Fwd(foo.Substring(1))|}.Substring(2))|}",
+                    @"C.Consume(C.Fwd(foo.AsSpan(1)).AsSpan(2))",
+                    new[] { 0, 1 },
+                    2, 1, 1
+                };
+                yield return new object[]
+                {
+                    receiver,
+                    @"{|#0:C.Consume({|#1:C.Fwd(foo.Substring(1), foo.Substring(2))|}.Substring(3), foo.Substring(4))|}",
+                    @"C.Consume(C.Fwd(foo.AsSpan(1), foo.AsSpan(2)).AsSpan(3), foo.AsSpan(4))",
+                    new[] { 0, 1 },
+                    2, 1, 1
+                };
+                yield return new object[]
+                {
+                    receiver,
+                    @"{|#0:C.Consume({|#1:C.Fwd(foo.Substring(1), {|#2:C.Fwd(foo.Substring(2))|}.Substring(3))|}.Substring(4))|}",
+                    @"C.Consume(C.Fwd(foo.AsSpan(1), C.Fwd(foo.AsSpan(2)).AsSpan(3)).AsSpan(4))",
+                    new[] { 0, 1, 2 },
+                    3, 1, 1
+                };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(Data_NestedViolations))]
+        public Task NestedViolations_AreAllReportedAndFixed_CS(
+            string receiverClass, string testExpression, string fixedExpression, int[] locations,
+            int? incrementalIterations = null, int? fixAllInDocumentIterations = null, int? fixAllIterations = null)
```

Comment:
All the yield return values in `Data_NestedViolations` seems to provide non-null values for these 3 parameters. Can these parameters be marked non-optional?

Reply:
Also it seems `fixAllInDocumentIterations` and `fixAllIterations` is always 1. Do we need a parameter for these as all?

Reply:
I think at one point I had different values for these parameters. I'll mark them non-optional and remove `fixAllInDocumentIterations` and `fixAllIterations`, as they are always 1.